### ISSUE TITLE
The option for fully baked lights

### ISSUE THREAD

**Issue by zomby138** [godotengine/godot-proposals#8040]
### Describe the project you are working on

Retro FPS with a lot of baked light, where performance is important.

### Describe the problem or limitation you are having in your project

It's not currently possible to make a light that is fully baked.  No matter what you do, the direct light is still calculated in real time on dynamic objects.  The direct light can be baked onto the lightmaps, but not onto the probes.  With shadows on, this light is very expensive, and even with shadows off, it costs more performance than it needs to.

Sometimes you want to use a light to make a room a little lighter, like a local ambient light.  For this it is very useful to be able to make a light that is 100% baked.  

Also, some older games did all of their lighting via something like probes, and we should be able to make games that emulate that retro style.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

You could add a thousand fully baked lights to a scene without it costing any runtime performance whatsoever (because the overhead from using lightmaps and probes is constant, no matter how many, or few lights you have in your scene.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

There would be one more option in the "Bake Mode" drop down for a light.  It could be called "Fully Baked", "Baked Direct", "Baked Only" or something.

A light with this setting would bake it's direct light into the like probes, in addition to the lightmap.  Then at runtime the light node does absolutely nothing whatsoever.  No being added to any clusteres, no calls to "void light()" in the shader of dynamic objects or anything.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't think there is any way to work around this.  If there it, I would be happy to do it.

### Is there a reason why this should be core and not an add-on in the asset library?

It's core in other engines.

**Comment by Calinou** [comment]
Reposting my [comment](https://github.com/godotengine/godot/issues/73289#issuecomment-1738332340) from https://github.com/godotengine/godot/issues/73289 for posterity:

___

I have a proof of concept: https://github.com/Calinou/godot/tree/light3d-add-static-baked-mode

**Testing project:** [test_light3d_baked_static_mode.zip](https://github.com/godotengine/godot/files/12744731/test_light3d_baked_static_mode.zip)

It has some issues:

- It crashes when you use **Scene > Reload Saved Scene** due to the culling code having some "uneven" cases with lights sometimes being paired, sometimes not. The goal here is to have the Static Baked lights *never* added to the list of instances to render. Only the lightmapper (and perhaps the VoxelGI baker in the future) should bother about these nodes.
- Toggling visibility of a Static Baked light breaks lightmap rendering, likely due to an error in my culling code changes.
- Lightmap probes don't take direct light into account for these Static Baked lights yet.

However, I was able to get a successful bake with lightmaps having the Static Baked lights present, yet dynamic objects not being lit by these lights in real-time at all.

Light3D bake modes from left to right: Disabled, Static, Dynamic, Static Baked. The white box is a static (baked) object, while the cyan torus is a dynamic object.

![image](https://github.com/godotengine/godot/assets/180032/04650a11-73aa-42ae-9e70-cc79184a66b8)

I'd appreciate help on this area as I'm not sure how to modify the lightmap probe baking code to get it to bake direct light for lights that use the Static Baked bake mode (and only those lights).

**Comment by jcostello** [comment]
@Calinou maybe Dario can help you since he worked in the implementation of the indirect light for probes

**Comment by zomby138** [comment]
@Calinou I'd like to get your branch running.  I built the code.  How do I set a light to be the new Static Baked mode?

**Comment by Calinou** [comment]
> @Calinou I'd like to get your branch running. I built the code. How do I set a light to be the new Static Baked mode?

There's a Bake Mode property (or GI Mode, I forgot) in the Light3D inspector. Note that the branch isn't in a fully functional state yet, so you'll need to modify to be able to use it in production.

**Comment by bertodelrio256** [comment]
Id like to add that it would also be nice to have a GI mode which allows a dynamic mesh to contribute to the lightmap "cast shadows, bounce light, etc.." but does not use the lightmap texture and uses the baked light probes. an example would be tree foliage, where I dont have enough lightmap density to cover an entire forest, but would like the leaves to cast shadows into the lightmap and just use probes for lighting.

**Comment by Arenovas** [comment]
I attempted to get something like this working by brute-forcing the lightprobes to just gather light from the first non bounced rays of light but I don't think I'm skilled enough to do this yet.
But I feel like this method should work unless I'm missing something critical, I have very little knowledge on how the entire light baking system works.
It would be really nice to see this kind of feature implemented though for both optimization reasons and for getting a more retro lighting feel in games that desire such.

**Comment by zomby138** [comment]
@Arenovas Hello!  Great to hear that someone else is also trying to get this working.  Together, I hope we can prevail.

I couldn't even find the relevant part of the code where the light baking happens.  I've messed around with spherical harmonics before and I feel like I could get it working if I knew how or where the baking happens right now.

**Comment by Arenovas** [comment]
Heya! And I really hope so!
I'm still not fully sure where or how the light baking works but I might've narrowed it down? I think lightmapper_rd.cpp does the calculations of where to place probes or which ones to use? And lm_compute.glsl is where the actual light calculations are handled, being called from the previously mentioned file.
In that compute shader there's a trace_direct_light function and also a trace_indirect_light function. The first most likely being used for the initial shadow pass and then the second being used for all the bounced lighting. So I'm guessing if the light probes could be forced to use the trace_direct_light function then maybe it'd work?
Of course that'd disable indirect lighting on them but it'd at least be a step in the right direction.

**Comment by zomby138** [comment]
@Arenovas  I've hacked something in.

<img width="1359" height="1091" alt="Image" src="https://github.com/user-attachments/assets/c22f7432-bee7-44fe-80e2-8e2df43e07c5" />

I've turned Indirect Energy down to 0.0 on this light so I can see if the direct is working.  I guess technically this is baking both direct and indirect light into the probes, but it's not tied to any sensible flags and the code's a bit of an ugly hack for now.

Still great progress!  I'm going to bed now.

**Comment by Arenovas** [comment]
Oooooo, very nice! And I mean getting together a hacky solution is a good start and was what I was aiming for myself just to figure out if it was possible.
Also I'd imagine in the end both direct and indirect light would be baked together into the probes, although I wonder if an option for just direct light could be handy too.

**Comment by zomby138** [comment]
Ok I've cleaned up the code and tested all the combinations I can think of.  Point lights, Spot lights, Directional lights.  Right now it does not honor the shadow setting of a light.  It always occludes the light.  But as far as I can tell this is the same as the behavior for the light baked on the walls, there are always shadows, no matter if the light has shadows set or not.

@Arenovas What is your use case for just baking the direct light into the probes without indirect?  Can you achieve this just by setting Indirect Energy to 0.0 on your lights?

What is the next step?  Right now, I have it baking direct light from any light that has "static_bake" set.  This is obviously wrong, it'll need to be tied to the new "Static Baked" Bake Mode.

Edit:  I've also noticed a problem.  It seems that the way it decides which probes to throw away isn't quite right for this.  You can see in the screen shot I posted before that the probes around the light have been discarded, even though if they existed, the change in light between them would be extreme.  Honestly, I could live with this for my use cases, but it is probably technically wrong.

**Comment by Arenovas** [comment]
Heck yeah! And yeah I'd imagine the shadowing toggle should probably match the baked texture. But I do wonder if that should be considered a bug with light baking overall or not.
And I was just imagining something like full on original Quake lighting which I don't think had indirect light? But yeah disabling indirect light on the lights themselves would probably achieve that.
Also tbh, the light probe system probably needs a major update at some point, considering it likes to generate them clipping into geometry and baking them pure black which you can also see in your screenshot.
I feel like something similar to the Magic Light Probes plugin from Unity could be a good system, generating them in the rough shape of the lighting and the average brightness/color. But that'd probably be a bit out of scope for this heheh.

**Comment by zomby138** [comment]
@Arenovas Yes most of that is beyond the scope of this proposal.  I want to get this one thing sorted, then we can move on to other tasks.  There is another lightprobe related feature that I want, but that is for later.

So what's the next step?  @Calinou has his proof of concept branch, but that's a couple of years out of date now, right?  Do we need to bring the branch up to date and then merge my code in to that?

I'll be honest, I haven't contributed to an open source project before.  I have no idea how the process works.

**Comment by Arenovas** [comment]
Yeah it's using Godot 4.2 or something as it's base so it'd definitely need updating. I'm not sure what the best way to go about doing so, my first thought is to just, port all of the changes by hand to the latest version but there's bound to be a better way.
But also I wonder if it'd be better to more use that branch as a reference instead of directly using the changes? Mostly since they mention some issues with culling changes they had made, dunno.

**Comment by Calinou** [comment]
> And yeah I'd imagine the shadowing toggle should probably match the baked texture. But I do wonder if that should be considered a bug with light baking overall or not.

This is due to https://github.com/godotengine/godot/issues/56611.

> And I was just imagining something like full on original Quake lighting which I don't think had indirect light? But yeah disabling indirect light on the lights themselves would probably achieve that.

Setting Indirect Energy to `0.0` on each Light3D node or Bounce Indirect Energy to `0.0` on LightmapGI should do the trick.

> Also tbh, the light probe system probably needs a major update at some point, considering it likes to generate them clipping into geometry and baking them pure black which you can also see in your screenshot.

https://github.com/godotengine/godot/pull/83420 aims to address this, but it needs to be reworked to use the mesh visual data instead of collision data.

**Comment by zomby138** [comment]
Ok.  So we could make a new branch and port the changes over.  How easy is it to see a list of changes that were made in that branch?  Is this something that you can do?

I'll attach the version of lm_compute.glsl that I have modified.  I haven't touched any other file.  If you build with this version of the file then any light set to Bake Mode "Static" will also bake it's direct light into probes.  The only think that will need to be changed is line 1899:

if (lights.data[i].static_bake)

This will need to instead check the for the new setting.

[lm_compute.glsl.txt](https://github.com/user-attachments/files/26147126/lm_compute.glsl.txt)

I had to add .txt to the file name to be able to attach it.

**Comment by Arenovas** [comment]
I did see that draft for fixing light probe positions, I was wondering whether it could just generate mesh colliders for every mesh just for calculating the positions but I guess that'd just an an extra step compared to using the visual data directly.

And @zomby138 you can access them from branch itself and clicking the "2 commits ahead of" text.

**Comment by zomby138** [comment]
After looking though the changes in @Calinou 's proof of concept, it doesn't look like it'll be hard to re-create them in a new branch.  I'm away for a day or so but I'll take a look at making a start when I get back.

**Comment by zomby138** [comment]
Ok I've made a branch with my baking code and all of @Calinou 's proof of concept code ported in.

https://github.com/zomby138/godot/tree/Fully-baked-lights

Everything builds and works ok.  Now it's time to decide how to pass this new bake state into the light baker

<img width="469" height="420" alt="Image" src="https://github.com/user-attachments/assets/ffbdbf39-3604-4132-9fc1-724cb76badec" />

Either I can add a new uint32_t bool to store it, or use the existing static_bake variable as an int/enum with something like
0 = dynamic
1 = static
2 = static baked

I don't know what the preference would be in the Godot code base.  Personally I don't like the idea of having two variables to store what is actually one setting.  I'm happy go with whatever Calinou wants.

There's clearly problems with the culling code.  If I set a light to the new Static Baked, the light stops being rendered (correctly).  But if I then change it back to Dynamic the light does not come back on.  I need to close the scene and re-open it again for the light to start working again.

**Comment by Arenovas** [comment]
Ooooo, very nice! And yeah using the existing variable would probably make the most sense but gotta see I guess.
And I am curious if the culling changes are needed or not. Not really sure what they do so maybe they are needed.

**Comment by Calinou** [comment]
> Either I can add a new uint32_t bool to store it, or use the existing static_bake variable as an int/enum with something like

I would go with the enum you mentioned, since it avoids scenarios where two booleans are set to incompatible values ("dynamic" and "static baked" at the same time).

**Comment by zomby138** [comment]
Just like @Calinou did before.  Cubes are static, cyan rings are dynamic, lights from left to right are bake disabled, static, dynamic, static_baked.

<img width="1738" height="445" alt="Image" src="https://github.com/user-attachments/assets/731efe4c-dab2-4a90-9b67-51d7a3ca52a2" />

**Comment by zomby138** [comment]
I'm really struggling with the culling stuff.  Directional lights in particular go into different lists than other lights.  But anyway here are some shots.  Cubes are static, torus are dynamic, and the one directional light is:

Bake Disabled:

<img width="1830" height="844" alt="Image" src="https://github.com/user-attachments/assets/e5463b4d-fb29-4e9e-8afd-826036553c23" />

Static:

<img width="1837" height="847" alt="Image" src="https://github.com/user-attachments/assets/bcfd8758-380a-4b98-9f0e-840211064276" />

Dynamic:

<img width="1837" height="847" alt="Image" src="https://github.com/user-attachments/assets/af690b5d-096f-40b1-8299-c4829d408e7c" />

Static Baked:

<img width="1843" height="847" alt="Image" src="https://github.com/user-attachments/assets/8f0fa3ee-5a75-4933-b646-4d2b58881145" />

This final one, the new one, has absolutely no real time light rendering whatsoever.  And the doughnuts look ok!

**Comment by jcostello** [comment]
Do you have a PR to try?

**Comment by zomby138** [comment]
@jcostello It's too early for a PR because there are tons of bugs with the culling code right now, but you can try it out on my branch here:

https://github.com/zomby138/godot/tree/Fully-baked-lights

And please do try it!  Right now if you bake a direction light, you will need to manually hide the light after the bake.  It will also crash when you close a scene that has a light set to Static Baked, so get used to opening your project in recovery mode.

**Comment by Arenovas** [comment]
No clue if I'll be able to manage anything to help with the culling code but I'm gonna go and poke around at it a bit. 
Definitely doing some good work there though!

**Comment by Arenovas** [comment]
Poking around and seems part of the issue is on line 1979 in render_scene_cull.cpp.
p_instance->array_index and swapped_instance->array_index end up higher than the swap_with_index values I think and causes issues.
Not sure on the details though behind it all but I just wanted to point this out but I'll keep looking into it.

**Comment by Arenovas** [comment]
Further research, hiding and showing an omni light with static baked enabled causes a CRASH_BAD_UNSIGNED_INDEX exception to occur, which doesn't crash the engine immediately, but as the light gets hidden and shown, it'll eventually reach down to 0 and overflow causing a full crash.

**Comment by Arenovas** [comment]
I haven't had any luck on fixing it sadly but hopefully what I wrote up can help some!
Also I noticed that when Godot is hooked up to the debugger it never actually crashes but it does when it's run without the debugger attached.

**Comment by Arenovas** [comment]
One more thing but is unrelated to the culling issues, I noticed that there's a little spot of light that appears on the backside of objects.

<img width="1002" height="508" alt="Image" src="https://github.com/user-attachments/assets/adc4da9b-50c7-4fc2-b567-2d5f3c0e2915" />

It's very noticeable when the indirect light energy is set to 0 and environment lighting is disabled, but even when those are on you can still see it a bit. Although with more striking colors like a bright pink it becomes more noticeable.
When I set it back to use indirect light baking only, it doesn't have the same issue.
I've only tested this with a directional light so far.

**Comment by zomby138** [comment]
> One more thing but is unrelated to the culling issues, I noticed that there's a little spot of light that appears on the backside of objects.

This one I can answer straight away.  This is just because of how spherical harmonics work, and I expected it before I started this work because I've messed around with spherical harmonics in the past.  If you look at this image:

<img width="1269" height="732" alt="Image" src="https://github.com/user-attachments/assets/13036827-a6d0-420c-a2ed-031ed811b258" />

A spherical harmonic is just a combination of these things.  If you think it through you are bound to end up with artifacts just like you're seeing there.  This is the same idea as Jpeg artifacts

<img width="849" height="243" alt="Image" src="https://github.com/user-attachments/assets/18e4f872-d8ab-4471-b25e-59e35564a4ca" />

When you try to represent something very high contrasts with a waveform, you get these echos.

Basically, spherical harmonics are good for soft, non high contrast situations where some light is coming from many directions.  They are not so good for Chiaroscuro style lighting.  If you want Chiaroscuro then a much simpler vector field would actually be much better, sort of like a radiosity normal map approach like they used back in half life 2.

**Comment by Arenovas** [comment]
Ohhhh, huh. So even when using the original indirect only baking those artifacts still appear but they're just far less noticeable due to it being a more subtle effect compared to direct light?

**Comment by zomby138** [comment]
That's one way of thinking about it.

**Comment by Jamsers** [comment]
I was trying to set up Half Life style lighting in a mobile project in Godot 4.6, and just couldn't get it working right. I was downright punching the desk, tearing my hair out in frustration, thinking I was missing something obvious, come to find, in 2026, on Godot 4.6, you actually just straight up can't light games the way they did in 1998.

I have no words. Open source game dev is incredible sometimes.

Anyway, I'm throwing my hat in the ring, yes this is needed. I don't want to be too mean. But to speak perfectly frankly, it's kinda crazy I, or anyone for that matter, even has to say this is needed. We've literally been lighting games like this since the _fucking 90s_.

Please also consider how well this works with shadowmasking. Please also consider, if there's gonna be a new bake mode, making SDFGI work in a similar manner. ([#8090](https://github.com/godotengine/godot-proposals/issues/8098))

Concerns about how low quality the lighting becomes when using this approach are irrelevant. The reason this approach is opted for in the first place is because of performance constraints. Developers should have the option to make that tradeoff.

**Comment by Arenovas** [comment]
I really wanna try and get this further but I'm still not fully sure how the culling system works and I guess sort of what it's for, I mean I guess optimization I'd imagine but not sure if there's anything else to it.
And also how exactly to go about figuring out how and why an index value ends up getting set incorrectly when using the baked direct lighting.

**Comment by zomby138** [comment]
> I really wanna try and get this further but I'm still not fully sure how the culling system works and I guess sort of what it's for, I mean I guess optimization I'd imagine but not sure if there's anything else to it. And also how exactly to go about figuring out how and why an index value ends up getting set incorrectly when using the baked direct lighting.

I also want to get this thing finished.  There are some other, related engine things that I want to look at but I want this one sorted first.  But I've been making real progress on my actual game instead lately.

The culling code basically is just deciding which lists the lights needs to be put in.  I would imagine that for one of these static baked lights we want it to be treated the same as a light with it's visibility turned off, with the one exception of it being in the list that gets used by the lightmap baker.

Maybe I'll take another stab at it next week.

**Comment by zomby138** [comment]
I've updated my branch and taken another stab at the culling code.  I think it works now.  Please if everyone could test it fully that would be great!

https://github.com/zomby138/godot/tree/Fully-baked-lights

Some new area lights have just turned up and I think I will need to get them working with the new bake mode, but other than that we might be very near the end here I hope!

**Comment by zomby138** [comment]
<img width="1796" height="390" alt="Image" src="https://github.com/user-attachments/assets/2d552bcd-b611-4640-a35b-6aceda007680" />

I got the new area lights working alright.  Although I don't think the probe light is going to be super accurate for larger area lights, probe light never really is super accurate anyway.  The rightmost cube is the new static baked mode so that doughnut is lit purely by the probes.

I updated my branch again. @Arenovas @Calinou please give my branch a full test if you can.  @Jamsers  I don't know anything about SDFGI, can you test my branch and see if things work the way you'd expect?

**Comment by Arenovas** [comment]
Did a quick test and seems like it's working pretty well so far! I definitely gotta try it out with some more complex scenes though.

**Comment by Calinou** [comment]
I tested the branch linked above and it works well, although I've seen light pairing error messages occasionally being spammed:

```
ERROR: BUG, indexing did not unpair geometries from light.
   at: instance_set_base (./servers/rendering/renderer_scene_cull.cpp:632)
```

**Testing project:** [test_fully_baked.zip](https://github.com/user-attachments/files/27136700/test_fully_baked.zip)

## Benchmark

<details open>

<summary>PC specifications</summary>

- **CPU:** AMD Ryzen 9 9950X3D
- **GPU:** NVIDIA GeForce RTX 5090
- **RAM:** 64 GB (2×32 GB DDR5-6000 CL30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 43)
</details>

Using a release export template (`production=yes lto=full`) in fullscreen 3840×2160.

Renderer | Dynamic | Static | Static Baked
-|-|-|-
Forward+ | 1003 FPS | 1398 FPS | 2519 FPS
Mobile | 1884 FPS | 2376 FPS | 5121 FPS
Compatibility | 1267 FPS | 1626 FPS | 6241 FPS

With llvmpipe at the default window size (1152×648):

Renderer | Dynamic | Static | Static Baked
-|-|-|-
Mobile | 15 FPS | 17 FPS | 24 FPS

llvmpipe at the smallest possible window size (64×64) to ensure fillrate isn't limiting things:

Renderer | Dynamic | Static | Static Baked
-|-|-|-
Mobile | 46 FPS | 48 FPS | 126 FPS

Additional notes:

- The test project's geometry is very simple. In a project with more complex geometry, the performance gain will be more significant when lights have shadows enabled.

- On top of improving performance significantly, this sidesteps the light limits in Mobile and Compatibility. There is no more light occasionally flickering on the dynamic object due to being affected by too many lights at once.

- This likely benefits mobile platforms and XR the most, given the limited CPU power there and the fact two viewports need to be rendered for XR.

**Comment by zomby138** [comment]
Thank you @Calinou.  Do you know of any way to provoke the error message?  I've been messing around with it today but I have not been able to get that error.

Edit: As soon as I made this comment, I got the error.

**Comment by zomby138** [comment]
I have push another update to my branch and I actually think it's all correct now.

With the exception of SDFGI.  I have never used that mode and I'm quite sure this setting will not work as expected in the mode.  Is it important that this works right away?  Or do you think we can go ahead with this just working for regular light mapping for now?

**Comment by Calinou** [comment]
> With the exception of SDFGI. I have never used that mode and I'm quite sure this setting will not work as expected in the mode. Is it important that this works right away? Or do you think we can go ahead with this just working for regular light mapping for now?

I think it's fine if this only supports LightmapGI, as I expect it to be the primary use case.

**Comment by jcostello** [comment]
Does the sky stop working when static baked the directional light?

**Comment by zomby138** [comment]
@jcostello I have no idea.  Does it?  Should it?  I don't use that auto sky thing.  How does it work right now, and how would you like it to work?

**Comment by Calinou** [comment]
> Does the sky stop working when static baked the directional light?

Environment/sky lighting is always static, and always has both direct and indirect light baked.

**Comment by jcostello** [comment]
Yeah, currently with static baked sky doesn't work any more. And it should work

**Comment by zomby138** [comment]
> Yeah, currently with static baked sky doesn't work any more. And it should work

Ok I'll take a look at it.

**Comment by zomby138** [comment]
> Yeah, currently with static baked sky doesn't work any more. And it should work

I've pushed an update that fixes the sky sun in full baked mode.

**Comment by DiviDevo** [comment]
If I may ask in my unending naivety and inexperience, how does this interact with shadowmasks/was this considered in the implementation?

Asking since my main concern here is cheap direct light shadows at a distance (via probes), which are then replaced with real time shadows close up (similar to CS:GO).

In the proposal nor here is the word "shadowmask" mentioned once (or at least I couldnt find the info I was looking for), would anyone care to clarify? Pretty please 🤔

Also please merge I need this good work thank you very cool :)
(I had this as a comment on the PR originally but moved it here since I think it fits better)

**Comment by zomby138** [comment]
@DiviDevo I've never used shadowmask.  I'm not familiar with how it works.  So probably it's not going to work the way you're hoping with this.  Please download my branch and test it.  Tell me what you expect to happen, and what actually happens.  I don't even know if shadowmask is something that people would expect to use with this feature, that's how little I know about it.

I agree about getting it merged in soon.  I don't know how long these things usually take.  It's been nearly a month now waiting for review.

**Comment by Calinou** [comment]
Shadowmasks require the light to be casting real-time light. Therefore, the Static Baked mode can't work in conjunction with a shadowmask (just like Static).

If you want to achieve lighting like in CS:GO or CS2, enable shadowmasking in LightmapGI, use Dynamic for the DirectionalLight3D and Static Baked for all other lights. (CS2 also uses real-time positional lights in select locations, but these can be turned off in the settings.)

Keep in mind Godot only supports shadowmasking for DirectionalLight3D, not other light types. Source 1/2 also share this limitation.

**Comment by Jamsers** [comment]
For shadowmasking, the fallback in the distance beyond the real time shadow range should act like what this proposal is asking for. Real time shadows are supposed to seamlessly fall back and blend into baked shadows - that's one of the supposed purposes of shadowmasking in the first place - it allows you to set the real time shadow range much shorter than you normally could because there's baked fallback that makes it much less jarring.

That's how it worked in CS:GO, CS2, Portal 2, the COD 4 era Call of Dutys, and UE4. The current Godot behavior where the direct light isn't baked (for probes) is actually quite odd... I genuinely haven't used another engine that even has that as a mode in the first place, at least for lightmaps.

**Comment by Jamsers** [comment]
For reference, this is how shadowmasking currently works in Godot:

https://github.com/user-attachments/assets/663315c0-26bb-4032-a0ff-0965d7980698

The actual shadowmasking implementation itself is excellent - notice how the real time shadows blend seamlessly with the static baked shadow on the floor. But the dynamic objects stick out like a sore thumb and "glow in the dark" as they fall out of range of the realtime shadow range - because of the bewildering behavior of not baking direct shadowing into probes at all.

**Comment by DiviDevo** [comment]
@Jamsers pretty much explained/showed what I mean quite well.
To me, the current implementation for shadowmasks feels very inconsistent/incomplete since I can blend between the baked lightmap and real time lighting on static geometry, but to correctly light the dynamic objects I need to extend the range of the real-time shadows?

I personally wouldnt mind if you were to merge now as-is and if Shadowmasks were tackled later, especially if you view these as separate problems.
But I guess to put my thumb on the scale, I would really like for it to be tackled just... at all.
Shadowmasking would be my preferred way of doing lighting since it's an excellent visual-performance trade, but the way it is now feels, simply, incomplete (and a bit irritating).

**Comment by zomby138** [comment]
Ok, I get the idea.  It reminds me very much of a problem in Unity's dual lightmaps like 15 years ago, back when Unity used Beast lightmapper.

I can see how you might think that this proposal is related to that problem, but in actuality it has no implementation overlap at all.

I can definitely think my way though solving that problem and I'm sure I could do it.  But it would need to be a different proposal and PR.  Also, I'm not starting work on anything else for the engine until this one gets merged.  There's no point in me doing any of this work in the first place if it doesn't get merged in anyway.

Is there anything I can do to help the merge process along?  Should I keep syncing my branch with godot:master or leave it as it is?
